src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs

```diff
@@ -0,0 +1,212 @@
+            newMethodBody = (BoundBlock)collector.Visit(methodBody);
+            dynamicAnalysisSpans = spansBuilder.ToImmutableAndFree();
+        }
+
+        private InstrumentationInjectionRewriter(MethodSymbol method, ArrayBuilder<SourceSpan> spansBuilder, FieldSymbol payload, TypeCompilationState compilationState, DiagnosticBag diagnostics, DebugDocumentProvider debugDocumentProvider)
+        {
+            _method = method;
+            _spansBuilder = spansBuilder;
+            _payload = payload;
+            _compilationState = compilationState;
+            _diagnostics = diagnostics;
+            _debugDocumentProvider = debugDocumentProvider;
+        }
+
+        public override BoundNode Visit(BoundNode node)
+        {
+            BoundStatement statement = node as BoundStatement;
+            if (statement != null)
+            {
+                switch (statement.Kind)
+                {
+                    case BoundKind.SwitchSection:
+                    case BoundKind.SwitchLabel:
+                    case BoundKind.PatternSwitchSection:
+                    case BoundKind.PatternSwitchLabel:
+                    case BoundKind.CatchBlock:
+                    // A labeled statement or a sequence point can be ignored with respect to instrumentation in favor of the underlying statement.
+                    case BoundKind.SequencePoint:
+                    case BoundKind.SequencePointWithSpan:
+                    case BoundKind.LabeledStatement:
+                        break;
+                    default:
+                        return CollectDynamicAnalysis(base.Visit(node));
+                }
+            }
+
+            return base.Visit(node);
+        }
+        
+        private BoundNode CollectDynamicAnalysis(BoundNode node)
+        {
+            BoundStatement statement = node as BoundStatement;
+            if (statement != null)
+            {
+                return CollectDynamicAnalysis(statement);
+            }
+
+            return node;
+        }
+
+        private BoundNode CollectDynamicAnalysis(BoundStatement statement)
+        {
+            if (statement.WasCompilerGenerated)
+            {
+                return statement;
+            }
+
+            Location statementLocation = statement.Syntax.GetLocation();
+            FileLinePositionSpan lineSpan = statementLocation.GetMappedLineSpan();
+            string path = lineSpan.Path;
+            if (path.Length == 0)
+            {
+                path = statement.Syntax.SyntaxTree.FilePath;
+            }
+
+            int spansIndex = _spansBuilder.Count;
+            _spansBuilder.Add(new SourceSpan(_debugDocumentProvider.Invoke(path, ""), lineSpan.StartLinePosition.Line, lineSpan.StartLinePosition.Character, lineSpan.EndLinePosition.Line, lineSpan.EndLinePosition.Character));
+
+            // Generate "_payload[pointIndex] = true".
```

Comment:
We haven't yet designed interactions between dynamic analysis and debugging, but I agree with @AlekseyTs that injected code should be treated as part of the associated statement.